txselector/txselector.go

```diff
@@ -1121,6 +1125,94 @@ func filterFailedAtomicGroups(
 	return txsToProcess, filteredTxs
 }
 
+// filterInvalidAtomicGroups split the txs into the ones that can be porcessed
+// and the ones that can't because they belong to an AtomicGroup that is impossible to forge
+// due to missing or bad ordered txs
+func filterInvalidAtomicGroups(
+	txs []common.PoolL2Tx,
+) (txsToProcess []common.PoolL2Tx, filteredTxs []common.PoolL2Tx) {
+	// validate if an atomic group is valid in terms of requested txs
+	isValid := func(atomicGroup common.AtomicGroup) bool {
+		for i := 0; i < len(atomicGroup.Txs); i++ {
+			// Find requested tx
+			rqRelativePosition, err := api.RequestOffset2RelativePosition(atomicGroup.Txs[i].RqOffset)
+			if err != nil {
+				// TODO: err should be compliant with txSelectorError
+				return false
+			}
+			requestedPosition := i + rqRelativePosition
+			if requestedPosition > len(atomicGroup.Txs)-1 || requestedPosition < 0 {
+				// TODO: err should be compliant with txSelectorError
+				return false
+			}
+			requestedTx := atomicGroup.Txs[requestedPosition]
+			// Check if the requested tx match the Rq fields
+			if atomicGroup.Txs[i].RqFromIdx != requestedTx.FromIdx ||
+				atomicGroup.Txs[i].RqToIdx != requestedTx.ToIdx ||
+				atomicGroup.Txs[i].RqToEthAddr != requestedTx.ToEthAddr ||
+				atomicGroup.Txs[i].RqToBJJ != requestedTx.ToBJJ ||
+				atomicGroup.Txs[i].RqTokenID != requestedTx.TokenID ||
+				atomicGroup.Txs[i].RqFee != requestedTx.Fee ||
+				atomicGroup.Txs[i].RqNonce != requestedTx.Nonce {
+				// TODO: err should be compliant with txSelectorError
+				return false
+			}
+			// Check amount
+			if atomicGroup.Txs[i].RqAmount != nil && requestedTx.Amount != nil {
+				// If both are different to nil
+				if atomicGroup.Txs[i].RqAmount.Cmp(requestedTx.Amount) != 0 {
+					// They must have same value (to be valid)
+					return false
+				}
+			} else if atomicGroup.Txs[i].RqAmount != requestedTx.Amount {
+				// Else they must be both nil (to be valid)
+				return false
+			}
+		}
+		return true
+	}
+	// build atomic groups
+	for i := 0; i < len(txs); i++ {
+		atomicGroupID := txs[i].AtomicGroupID
+		if atomicGroupID == common.EmptyAtomicGroupID {
+			// Tx is not atomic, not filtering
+			txsToProcess = append(txsToProcess, txs[i])
+			continue
+		} else {
+			// Tx is atomic
+			// Create atomic group
+			atomicGroup := common.AtomicGroup{}
+			for i < len(txs) && txs[i].AtomicGroupID == atomicGroupID {
```

Comment:
In order to avoid this inner tx loop and reduce a bit the complexity of this code, what do you think about grouping the txs in a map where the key is the AtomicGroupID and the value is a slice of the tx that compose the AtomicGroup? 

With this, we could loop through the txs once to build the map and then loop through the map once to process all the transactions.

Reply:
This was actually my first attempt, but this was resulting in modifying the original order, which complicated the tests. So instead of fixing the test I did this, shame on me xD